arbitrator/jit/src/wavmio.rs

```diff
@@ -0,0 +1,311 @@
+}
+
+pub fn set_global_state_bytes32(env: &WasmEnvArc, sp: u32) -> MaybeEscape {
+    let (sp, mut env) = GoStack::new(sp, env);
+    ready_hostio(&mut *env)?;
+
+    let global = sp.read_u64(0) as u32 as usize;
+    let src_ptr = sp.read_u64(1);
+    let src_len = sp.read_u64(2);
+    if src_len != 32 {
+        eprintln!("Go trying to set 32-byte global with a {src_len} bytes long buffer");
+        return Ok(());
+    }
+
+    let slice = sp.read_slice(src_ptr, src_len);
+    let slice = &slice.try_into().unwrap();
+    match env.large_globals.get_mut(global) {
+        Some(global) => *global = *slice,
+        None => {
+            return Escape::hostio("global write out of bounds in wavmio.setGlobalStateBytes32")
+        }
+    }
+    Ok(())
+}
+
+pub fn get_global_state_u64(env: &WasmEnvArc, sp: u32) -> MaybeEscape {
+    let (sp, mut env) = GoStack::new(sp, env);
+    ready_hostio(&mut *env)?;
+
+    let global = sp.read_u64(0) as u32 as usize;
+    match env.small_globals.get(global) {
+        Some(global) => sp.write_u64(1, *global),
+        None => return Escape::hostio("global read out of bounds in wavmio.getGlobalStateU64"),
+    }
+    Ok(())
+}
+
+pub fn set_global_state_u64(env: &WasmEnvArc, sp: u32) -> MaybeEscape {
+    let (sp, mut env) = GoStack::new(sp, env);
+    ready_hostio(&mut *env)?;
+
+    let global = sp.read_u64(0) as u32 as usize;
+    match env.small_globals.get_mut(global) {
+        Some(global) => *global = sp.read_u64(1),
+        None => return Escape::hostio("global write out of bounds in wavmio.setGlobalStateU64"),
+    }
+    Ok(())
+}
+
+pub fn read_inbox_message(env: &WasmEnvArc, sp: u32) -> MaybeEscape {
+    let (sp, mut env) = GoStack::new(sp, env);
+    ready_hostio(&mut *env)?;
+
+    let inbox = &env.sequencer_messages;
+    inbox_message_impl(&sp, &env, inbox, "wavmio.readInboxMessage")
+}
+
+pub fn read_delayed_inbox_message(env: &WasmEnvArc, sp: u32) -> MaybeEscape {
+    let (sp, mut env) = GoStack::new(sp, env);
+    ready_hostio(&mut *env)?;
+
+    let inbox = &env.delayed_messages;
+    inbox_message_impl(&sp, &env, inbox, "wavmio.readDelayedInboxMessage")
+}
+
+/// Reads an inbox message
+/// note: the order of the checks is very important.
+fn inbox_message_impl(
+    sp: &GoStack,
+    env: &MutexGuard<WasmEnv>,
+    inbox: &Inbox,
+    name: &str,
+) -> MaybeEscape {
+    let msg_num = sp.read_u64(0);
+    let offset = sp.read_u64(1);
+    let out_ptr = sp.read_u64(2);
+    let out_len = sp.read_u64(3);
+    if out_len != 32 {
+        eprintln!("Go trying to read inbox message with out len {out_len} in {name}");
+        sp.write_u64(5, 0);
+        return Ok(());
+    }
+
+    macro_rules! error {
+        ($text:expr $(,$args:expr)*) => {{
+            let text = format!($text $(,$args)*);
+            return Escape::hostio(&text)
+        }};
+    }
+
+    let too_far = env.first_too_far;
+    let message = match inbox.get(&msg_num) {
+        Some(message) => message,
```

Comment:
Yes, `first_to_far` is the first sequencer inbox message that's out-of-bounds. This is also only used to make the error message more helpful since we don't actually have a concept of `TOO_FAR` in this validator. I've gone ahead though and collapsed these both down into a single error message and removed `first_to_far`.

I went ahead and fixed the Arbitrator edge cases we talked about on slack too